src/zkevm_specs/exp_circuit.py

```diff
@@ -0,0 +1,103 @@
+from typing import List
+from .evm import (
+    ExpCircuit,
+    ExpCircuitRow,
+)
+from .util import (
+    ConstraintSystem,
+    FQ,
+    RLC,
+    mul_add_words,
+    word_to_lo_hi,
+)
+
+
+def verify_step(cs: ConstraintSystem, rows: List[ExpCircuitRow]):
+    # for every step except the last
+    with cs.condition(rows[0].q_step * (1 - rows[0].is_last)) as cs:
+        # base is the same across rows.
+        cs.constrain_equal(rows[0].base, rows[1].base)
+        # multiplication result from the "next" row `d` must be used as
+        # the first multiplicand in the "cur" row `a`
+        cs.constrain_equal(rows[0].a, rows[1].d)
+        # identifier does not change over the steps of an exponentiation trace.
+        cs.constrain_equal(rows[0].identifier, rows[1].identifier)
+
+    # for every step
+    with cs.condition(rows[0].q_step) as cs:
+        # is_last is boolean.
+        cs.constrain_bool(rows[0].is_last)
+        # is_pad is boolean.
+        cs.constrain_bool(rows[0].is_pad)
+        # is_odd, i.e. odd/even parity is a boolean.
+        cs.constrain_bool(rows[0].is_odd)
+        # is_last == 1 is followed by is_pad == 1.
+        # is_last == 0 is following by is_pad == 0.
+        cs.constrain_equal(rows[0].is_last, rows[1].is_pad)
+        # multiplication is assigned correctly
+        _overflow, carry_lo_hi, additional_constraints = mul_add_words(
+            rows[0].a, rows[0].b, rows[0].c, rows[0].d
+        )
+        cs.range_check(carry_lo_hi[0], 9)
+        cs.range_check(carry_lo_hi[1], 9)
+        cs.constrain_equal(additional_constraints[0][0], additional_constraints[0][1])
+        cs.constrain_equal(additional_constraints[1][0], additional_constraints[1][1])
```

Comment:
We do the `mul_add_words` operation which returns 2 additional constraints that need to hold.

The util method simply returns those constraints, then I use the constraint system to validate those constraints:
1. `mul_add_words`: https://github.com/scroll-tech/zkevm-specs/blob/feat/exp/src/zkevm_specs/util/arithmetic.py#L135
2. additional constraints: https://github.com/scroll-tech/zkevm-specs/blob/feat/exp/src/zkevm_specs/util/arithmetic.py#L161-L162

For simplicity, I could have used `assert` in the util itself, but then in case of failure it wouldn't catch and report it appropriately (this part is managed by the [constraint system](https://github.com/scroll-tech/zkevm-specs/blob/feat/exp/src/zkevm_specs/util/constraint_system.py))